### PR TITLE
migrate to pkg web

### PR DESCRIPTION
**Pull request by kevmoo**

### CODE REVIEW

**Comment by kevmoo** [comment]
I'm not racing to publish this, since devtoool + wasm still doesn't load with this change.

But all of the JS tests pass and it still works with devtools + JS – so we're closer to where we want to be, I guess

File: lib/src/web_helpers.dart [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import 'package:web/helpers.dart';
+
+extension WebSocketExtension on WebSocket {
```
Comment (kevmoo) [inline lib/src/web_helpers.dart:3]: @srujzs – might could go into pkg:web?

File: lib/html.dart [hunk 2]
```diff
@@ -73,7 +75,8 @@ class HtmlWebSocketChannel extends StreamChannelMixin
   /// [BinaryType.blob], they're delivered as [Blob]s instead.
   HtmlWebSocketChannel.connect(Object url,
       {Iterable<String>? protocols, BinaryType? binaryType})
-      : this(WebSocket(url.toString(), protocols)
+      : this(WebSocket(url.toString(),
+            (protocols ?? <String>[]).toList().map((e) => e.toJS).toList().toJS)
```
Comment (srujzs) [inline lib/html.dart:79]: nit: move the first `toList` up to `protocols` e.g. `protocols?.toList()` so you don't have to recreate the list in the case where `protocols` is null.

File: pubspec.yaml [hunk 3]
```diff
@@ -14,6 +14,7 @@ dependencies:
   async: ^2.5.0
   crypto: ^3.0.0
   stream_channel: ^2.1.0
+  web: ^0.4.0
```
Comment (devoncarew) [inline pubspec.yaml:17]: Its definitely good to start dogfooding this stuff; does this in any way tie our hands wrt publishing packages that now depend on it?

File: lib/html.dart [hunk 4]
```diff
@@ -124,16 +123,27 @@ class HtmlWebSocketChannel extends StreamChannelMixin
     });
   }
 
+  void _innerListen(MessageEvent event) {
+    dynamic data = event.data;
```
Comment (srujzs) [inline lib/html.dart:127]: I don't understand why this has to be `dynamic`. `event.data` is a `JSAny?` to begin with, no?

File: lib/src/web_helpers.dart [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import 'package:web/helpers.dart';
```
Comment (devoncarew) [inline lib/src/web_helpers.dart:1]: DBC: this file needs a copyright header

**Review by srujzs — APPROVED** [review]

File: lib/html.dart (hunk 2 above)
Comment (kevmoo) [reply lib/html.dart]: yup. See what i did...

File: pubspec.yaml (hunk 3 above)
Comment (NotTsunami) [reply pubspec.yaml:17]: Might want to allow the same version range as http (from https://github.com/dart-lang/http/pull/1055) if a release is planned anytime soon (flutter 3.16.1 pins web 0.3.0)

File: lib/html.dart [hunk 6]
```diff
@@ -73,8 +75,15 @@ class HtmlWebSocketChannel extends StreamChannelMixin
...
-      : this(WebSocket(url.toString(), protocols)
-          ..binaryType = (binaryType ?? BinaryType.list).value);
+      : this(
+          WebSocket(
+            url.toString(),
+            (protocols?.toList() ?? const <String>[])
+                .map((e) => e.toJS)
+                .toList()
+                .toJS,
+          )..binaryType = (binaryType ?? BinaryType.list).value,
```
Comment (felangel) [inline lib/html.dart:85]: Why is the default `arraybuffer`? Looks like it's `blob` according to https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/binaryType

**Pull request merged**